alpa/create_state_parallel.py

```diff
@@ -0,0 +1,177 @@
+"""Compile executables for creating training state distributedly."""
+from collections import defaultdict, deque
+from typing import Sequence, Optional
+
+from jax._src.lib import xla_bridge as xb, xla_extension as xe
+from jax.core import (Jaxpr, ClosedJaxpr, Literal, new_jaxpr_eqn, gensym,
+                      get_aval, raise_to_shaped, AbstractValue, Var)
+from jax.interpreters import partial_eval as pe, pxla
+from jax.tree_util import tree_flatten, tree_unflatten, PyTreeDef
+
+from alpa.device_mesh import ReplicatedDistributedArray, PhysicalDeviceMeshGroup
+from alpa.measure_record import StrategyConfig
+from alpa.mesh_executable import NormalMeshDriverExecutable, GradAccMeshDriverExecutable, PlacementSpec
+from alpa.pipeline_parallel.compile_executable import compile_pipeshard_executable_internal
+from alpa.pipeline_parallel.layer_construction import add_pipeline_marks_for_sliced_eqns
+from alpa.pipeline_parallel.pipeshard_executable import PipeshardDriverExecutable
+from alpa.pipeline_parallel.runtime_emitter import PipeshardConfig
+from alpa.pipeline_parallel.stage_construction import UniformStageOption
+from alpa.shard_parallel.auto_sharding import (run_auto_sharding_pass, AutoShardingOption,
+    run_spmd_partitioner_pass)
+from alpa.util import jaxpr_to_hlo_module, OrderedSet
+
+
+class CreateStateExecutable(PipeshardDriverExecutable):
+    """
+    A distributed executable that creates a training state for a function
+    parallelized by PipeshardParallel.
+    """
+    def __init__(self,
+                 mesh_group: PhysicalDeviceMeshGroup,
+                 pipeshard_config: PipeshardConfig,
+                 placement_specs: Sequence[PlacementSpec],
+                 out_tree: Optional[PyTreeDef] = None,
+                 static_argnums: Optional[Sequence[int]] = None):
+        super().__init__(mesh_group=mesh_group,
+                         pipeshard_config=pipeshard_config,
+                         num_batch=1,
+                         out_tree=out_tree,
+                         static_argnums=static_argnums)
+        self.placement_specs = placement_specs
```

Comment:
I suggest we use different names for input and output placement specs. Like `get_placement_specs`->`get_input_placement_specs`. Otherwise this class has a `placement_specs` member and a `get_placement_specs` method(inherited from PipeshardParallel) but the two have different meaning

Reply:
Sounds good. I renamed `get_placement_specs` to `get_input_placement_specs` and renamed `placement_specs` in `CreateStateExecutable` to `target_placement_specs`